src/drivers/intel/ssp/ssp.c

```diff
@@ -742,6 +787,78 @@ static int ssp_get_hw_params(struct dai *dai,
 	return 0;
 }
 
+/*
+ * enabled clock MCLK/BCLK to support codec that need MCLK/BCLK
+ * This is called before trigger(COMP_TRIGGER_START)
+ */
+static int ssp_hw_params(struct dai *dai,
+			 struct sof_ipc_stream_params  *params)
+{
+	struct ssp_pdata *ssp = dai_get_drvdata(dai);
+	int ret;
+
+	if (ssp->state[SOF_IPC_STREAM_CAPTURE] < COMP_STATE_PREPARE &&
+	    ssp->state[SOF_IPC_STREAM_PLAYBACK] < COMP_STATE_PREPARE) {
+		dai_err(dai, "ssp_hw_params(): Please set config first");
+		return -EINVAL;
+	}
+
+	dai_info(dai, "ssp_hw_params() params.clks_control 0x%x", ssp->params.clks_control);
```

Comment:
We should not set mclk/bclk in case already set, do we need some similar (but not same) logic as below?
```
	/* SSP active means bclk already configured. */
	if (ssp->state[SOF_IPC_STREAM_PLAYBACK] == COMP_STATE_ACTIVE ||
	    ssp->state[SOF_IPC_STREAM_CAPTURE] == COMP_STATE_ACTIVE)
		return 0;
```

Reply:
> We should not set mclk/bclk in case already set, do we need some similar (but not same) logic as below?
> 
> ```
> 	/* SSP active means bclk already configured. */
> 	if (ssp->state[SOF_IPC_STREAM_PLAYBACK] == COMP_STATE_ACTIVE ||
> 	    ssp->state[SOF_IPC_STREAM_CAPTURE] == COMP_STATE_ACTIVE)
> 		return 0;
> ```

@keyonjie Thanks for pointing it out, but we may not be able to use `ssp->state == COMP_STATE_ACTIVE` to test if the clock is set. We don't change ssp->state when the clock is set in hw_params. Maybe we need to add a flag to identify the clock status?

Reply:
@keyonjie updated.